<a href="https://colab.research.google.com/github/sophie55555/Project/blob/main/project3_load_breast_cancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

6-3. 프로젝트 (3) load_breast_cancer : 유방암 여부를 진단해 봅시다
마지막으로 유방암 여부를 진단해 보겠습니다.
이 데이터 또한 여러 사람의 건강 지표에 대한 데이터가 feature로 들어가있고, 유방암의 여부가 True, False로 label이 됩니다.

주어진 데이터로 환자의 유방암 여부를 분류해 볼 수 있을까요?

(1) 필요한 모듈 import하기

In [10]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import pandas as pd
import numpy as np

(2) 데이터 준비

In [3]:
cancer = load_breast_cancer() # Dictionary 자료형과 유사한 sklearn.utils.Bunch 자료형
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

(3) 데이터 이해하기
지피지기면 백전불태! 다루어야 할 데이터를 자세히 살펴봅시다.

Feature Data 지정하기
Label Data 지정하기
Target Names 출력해 보기
데이터 Describe 해 보기

In [4]:
cancer_data = cancer.data
cancer_data.shape

(569, 30)

In [5]:
# Feature 데이터 추출
X = cancer.data
print("Feature 데이터 (X):\n", X[:5]) # 처음 5개 행 출력
print("Feature 데이터 형태:", X.shape) # Feature 데이터의 형태 (행, 열) 출력

Feature 데이터 (X):
 [[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
  1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
  6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
  1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
  4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 1.326e+03 8.474e-02 7.864e-02 8.690e-02
  7.017e-02 1.812e-01 5.667e-02 5.435e-01 7.339e-01 3.398e+00 7.408e+01
  5.225e-03 1.308e-02 1.860e-02 1.340e-02 1.389e-02 3.532e-03 2.499e+01
  2.341e+01 1.588e+02 1.956e+03 1.238e-01 1.866e-01 2.416e-01 1.860e-01
  2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 1.203e+03 1.096e-01 1.599e-01 1.974e-01
  1.279e-01 2.069e-01 5.999e-02 7.456e-01 7.869e-01 4.585e+00 9.403e+01
  6.150e-03 4.006e-02 3.832e-02 2.058e-02 2.250e-02 4.571e-03 2.357e+01
  2.553e+01 1.525e+02 1.709e+03 1.444e-01 4.245e-01 4.504e-01 2.430e-01
  3.613e-01 8.758e-02]
 [1.142e+01 2.038e+01 7.758e+01 3.861e+02 1.425e-

In [6]:
# Feature 이름 추출
feature_names = cancer.feature_names
print("\nFeature 이름:\n", feature_names)


Feature 이름:
 ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


In [7]:
# Label 데이터 추출
y = cancer.target
print("\nLabel 데이터 (y):\n", y[:5]) # 처음 5개 레이블 출력
print("Label 데이터 형태:", y.shape) # Label 데이터의 형태 출력


Label 데이터 (y):
 [0 0 0 0 0]
Label 데이터 형태: (569,)


In [8]:
# Label 이름 추출
target_names = cancer.target_names
print("\nLabel 이름:\n", target_names)


Label 이름:
 ['malignant' 'benign']


In [11]:
# DataFrame 생성 (데이터 확인 및 분석 용이)
data = pd.DataFrame(cancer.data, columns=cancer.feature_names)
target = pd.DataFrame(cancer.target, columns=['target'])
df = pd.concat([data, target], axis=1)

In [12]:
# Feature와 Label 분리
X = cancer.data
y = cancer.target

(4) train, test 데이터 분리

모델 학습과 테스트용 문제지와 정답지를 준비해 봅시다. X_train, X_test, y_train, y_test를 생성하는 방법을 참고해 보세요.

In [13]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
'''
sklearn.model_selection.train_test_split(*arrays, test_size=None, train_size=None, random_state=None, shuffle=True, stratify=None)


# test_size : 전체 데이터 셋 중 Test Dataset에 포함하려고 하는 비율
# shuffle : Training Dataset과 Test Dataset으로 나누기 전에, 데이터들을 뒤섞을지 여부
'''
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15)

decision_tree = DecisionTreeClassifier(random_state=15)
decision_tree.fit(X_train, y_train) # 의사결정나무 모델로 학습
y_pred = decision_tree.predict(X_test) # 테스트 결과 예측

accuracy = accuracy_score(y_test, y_pred) # y_pred(답안지)를 y_test(정답지)로 채점
accuracy # 정확도 출력

0.9473684210526315

RandomForest

In [15]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_iris #싸이킷런 데이터셋에 있는 iris를 불러오는 코드
from sklearn.model_selection import train_test_split #싸이킷런에 있는 데이터를 train과 test dataset으로 나누어주는 함수 import
from sklearn.ensemble import RandomForestClassifier #랜덤포레스트라는 분류기를 사용하기 위해 import
from sklearn.metrics import classification_report #싸이킷런에 있는 분류 결과에 대한 시각화를 위해 쓰는 코드

# (2) 데이터 준비
cancer_data = cancer.data #iris데이터의 data컬럼을 분류해 iris_data 변수에 담는다.
cancer_label = cancer.target #iris데이터의 target컬럼을 분류해 iris_label 변수에 담는다.


X_train, X_test, y_train, y_test = train_test_split(cancer_data, # iris 데이터의 data 컬럼
                                                    cancer_label, # iris 데이터의 target 컬럼
                                                    test_size=0.2, # test_size : train data와 test data를 몇대몇으로 나눌지 정하는 옵션
                                                    random_state=21) # random_state : 랜덤 패턴의 값을 지정

random_forest = RandomForestClassifier(random_state=32) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

              precision    recall  f1-score   support

           0       0.97      0.95      0.96        39
           1       0.97      0.99      0.98        75

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114



SVM 모델은 다음과 같이 사용합니다.

In [16]:
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

print(svm_model._estimator_type) # 이 모델이 어떤 타입인지 확인한다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       0.93      0.72      0.81        39
           1       0.87      0.97      0.92        75

    accuracy                           0.89       114
   macro avg       0.90      0.85      0.86       114
weighted avg       0.89      0.89      0.88       114



SGD Classifier 모델은 다음과 같이 사용합니다.

In [17]:
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

print(sgd_model._estimator_type) # 이 모델의 타입을 확인

sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       0.89      0.79      0.84        39
           1       0.90      0.95      0.92        75

    accuracy                           0.89       114
   macro avg       0.89      0.87      0.88       114
weighted avg       0.89      0.89      0.89       114



Logistic Regression 모델은 다음과 같이 사용합니다.

In [18]:
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression() # 모델 객체 생성

print(logistic_model._estimator_type) # 이 모델의 타입을 확인

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       0.92      0.85      0.88        39
           1       0.92      0.96      0.94        75

    accuracy                           0.92       114
   macro avg       0.92      0.90      0.91       114
weighted avg       0.92      0.92      0.92       114



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(6) 모델을 평가해 보기 학습된 모델들의 테스트데이터 예측 결과를 어떻게 해석해야 할까요? 모델의 성능을 평가하는 지표로는 무엇이 좋을까요? sklearn.metrics 에서 제공하는 평가지표 중 적절한 것을 선택해 보세요. 선택하신 이유도 설명해 주세요.

In [19]:
fake_pred = [0] * len(y_pred) # 0으로만 이루어진 가짜 답안지 준비

accuracy = accuracy_score(y_test, fake_pred) # fake_pred(가짜 답안지)를 y_test(정답지)로 채점
accuracy # 정확도 출력

0.34210526315789475

In [20]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[33,  6],
       [ 3, 72]])

In [21]:
confusion_matrix(y_test, fake_pred)

array([[39,  0],
       [75,  0]])

In [22]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.85      0.88        39
           1       0.92      0.96      0.94        75

    accuracy                           0.92       114
   macro avg       0.92      0.90      0.91       114
weighted avg       0.92      0.92      0.92       114



In [23]:
print(classification_report(y_test, fake_pred, zero_division=0))

              precision    recall  f1-score   support

           0       0.34      1.00      0.51        39
           1       0.00      0.00      0.00        75

    accuracy                           0.34       114
   macro avg       0.17      0.50      0.25       114
weighted avg       0.12      0.34      0.17       114



In [24]:
accuracy_score(y_test, y_pred), accuracy_score(y_test, fake_pred)

(0.9210526315789473, 0.34210526315789475)

(모델 평가)

RandomForest - 97%

RandomForest - 평균 97%의 높은 정확도를 보임.

SVM - 89%

SGD Classifier - 클래스별 성능 차이:
클래스 0 (악성): 정밀도는 0.89로 비교적 높지만, 재현율은 0.79로 상대적으로 낮음. 이는 모델이 악성이라고 예측한 것 중 실제 악성인 비율은 높지만, 실제 악성 샘플 중 모델이 악성이라고 제대로 찾아낸 비율은 비교적 낮다는 것을 의미함. 즉, 실제 악성인데 양성으로 잘못 분류하는 경우가 있을 수 있음.
클래스 1 (양성): 정밀도는 0.90, 재현율은 0.95로 모두 높음. 양성 샘플을 비교적 정확하게 분류하고 있음을 나타냄.
클래스 불균형: 양성(75개) 샘플이 악성(39개) 샘플보다 많음. 이러한 클래스 불균형이 모델 성능에 영향을 미쳤을 가능성이 있으며, 특히 악성 샘플의 재현율이 낮은 것에 영향을 주었을 수 있음.

Logistic Regression - 클래스 0 (악성): 정밀도는 0.92로 높지만, 재현율은 0.85로 상대적으로 낮음. 즉, 악성이라고 예측한 것 중 실제 악성인 비율은 높지만, 실제 악성 샘플 중 일부를 놓치는 경우가 있다는 것을 의미함.
클래스 1 (양성): 정밀도와 재현율이 각각 0.92와 0.96으로 모두 높음. 양성 샘플을 매우 잘 분류하고 있음을 나타냄.

accuracy_score(y_test, y_pred): 모델의 정확도
accuracy_score(y_test, fake_pred): 가짜 예측의 정확도

두 정확도의 차이(92.1% vs 34.2%)는 모델이 학습을 통해 얼마나 효과적으로 데이터를 학습했는지, 즉 의미 있는 패턴을 얼마나 잘 찾아냈는지를 명확하게 보여줍니다.